박스오피스 [STEP 3] Volga

volgaHwang opened the pull request:
@zdodev 

안녕하세요! 소대!

설 연휴는 잘 쉬셨나요?

오늘은,STEP3 PR을 요청 드립니다!

## 고민한 점

이번 스텝도 역시 고민할 거리가 많았습니다..!!

- 최대한 파일을 많이 쪼개기 했습니다.
- Modern CollectionView Implementation을 보고 Layout / Section 구현법을 보고 참조를 많이 했습니다.
- Network 통신 이후 받아온 데이터를 UI에 반영하기 위해 `updateUI()` 메소드를 구현했습니다.
- `BoxOfficeListCell`에 매핑할 `BoxOfficeItem` `Entity`를 생성했습니다.
- `BoxOfficeItem`은 받아오는 `DailyBoxOffice` Modeld을 바로 매핑할수 있게 `Designated Initializer`를 구현했습니다.
- 날짜를 받아오는 포맷이 `Title`과 어제 BoxOfficeList를 받아오기 위한 `Query`에 들어가는 것이 다른 부분에 대해 `Date` Class에 `dayString` 타입 메소드를 구현했고, format이나 날짜에 대한 부분에서는 Enumerate를 추가하여 구현했습니다.
- `Cell`에서 AutoLayout에 의한 리스크를 줄이고자 StackView를 많이 사용 했습니다.

## 궁금한 점

- `refreshControl` 구현한 부분이 있는데, `loadData` 이후에 `endRefreshing()`메소드를 바로 호출 해 주는데, 이 부분이 UX상 맞는지 잘 모르겠습니다.
- 아직까지는? `BoxOfficeListViewController`에 코드가 감당 안될 정도로 많은편은 아니지만 조만간 더 많아질 것 같다는 생각이 드는 구조인데, 이 구조에 확신이 없는것같습니다. 혹시 더 컴팩트하게 줄일 방법이 있을까요?
- `CollectionView`에서 ViewController에 정의한 `createLayout()`을 사용하기 위해서 `lazy var`키워드를 사용했는데 이것도 맞게 사용 한 것인지 잘 모르겠습니다..! 더 좋은 방법이 있을까요?

volgaHwang committed:
chore: Modern CollectionView 사용을 위한 버전 변경 반영

- CellRegistration + UICollectionViewDiffableDataSource Provider 사용을 위한 버전 변경사항 반영

volgaHwang committed:
chore: Scene Directory 생성 반영

volgaHwang committed:
refactor: ViewController -> BoxOfficeListViewController 리네이밍 반영

volgaHwang committed:
feat: BoxOfficeListViewController 네비게이션으로 감싸기

volgaHwang committed:
feat: BoxOfficeListViewController Section 추가사항 반영

volgaHwang committed:
feat: CollectionView 추가

- List로 사용할 CollectionView를 추가했습니다.
- createLayout 메소드를 사용해 UICollcetionView 초기화에 사용할 layout을 구현합니다.
- ViewController 내부 메소드를 사용하기 위해 collectionView를 `lazy var` 키워드를 사용해서 선언했습니다.

volgaHwang committed:
feat: BoxOfficeListViewController 계층 구조 반영

volgaHwang committed:
feat: BoxOfficeListCell 추가 사항반영

- 디렉토리와 파일을 추가하였습니다.

volgaHwang committed:
feat: BoxOfficeListCell 내부 구조 구현

- `accessoryImageView`를 따로 구현했습니다.
- separator를 `UIView()`를 통해 구현했습니다.
- 전체적인 틀은 `UIStackView`를 사용하여 구조를 잡아두었습니다.
- 초기화 시 `configure()` 메소드를 통해 조절하게 구현했습니다.

volgaHwang committed:
feat: BoxOfficeItem 구조 반영

volgaHwang committed:
feat: BoxOfficeItem 구현사항 반영

- BoxOffice Item Property를 크게 5가지로 정의했습니다.
    - `title`: 영화 제목
    - `showCountInformation`: 일일 관객 / 총 관객 에 대한 정보를 담고 있는 String 구조
    - `rank`: 박스 오피스 순위에 대한 정보를 담는 프로퍼티
    - `rankDescription`: 랭크 순위에 대해서 변동사항 또는 신작일 경우에 대한 Description 글자 색상 변경사항이 있으므로 NSAttributedString

volgaHwang committed:
feat: BoxOfficeItem 초기화 및 String Extension 추가사항 반영

- BoxOfficeItem은 CollectionView Cell에 들어갈 정보를 담고 있는 구조체로 사용합니다.
- DailyBoxOffice 에 대해 매핑을 바로 할수 있는 initializer를 구현했습니다.
- String Extension에 대한 부분은 해당 파일 내에서만 사용되므로 private Extension을 통해 구현했습니다.

volgaHwang committed:
feat: BoxOfficeItem 추가사항 반영

- BoxOfficeContents Enumerate 추가하여, 상수 부분 처리하였습니다.

volgaHwang committed:
feat: BoxOfficeListViewController DataSource Configuration 부분 추가 반영

volgaHwang committed:
feat: Date: yesterday Type Property 추가

volgaHwang committed:
feat: BoxOfficeListViewController 구현사항 추가

- networkService 프로퍼티 구하해서 네트워크 사용하게 반영
- loadData 메소드를 통해 네트워킹
- updateUI 통해, CollectionView Cell 업데이트 트리거

volgaHwang committed:
feat: BoxOfficeViewController 속성 변경사항 반영하는 함수 추가 + 컴파일 패스 변경

volgaHwang committed:
feat: BoxOfficeListViewController: refresh Control 추가사항 반영

volgaHwang committed:
refactor: Date 날짜 가져오는 변경사항 반영

- 기존에 static var를 통해 today / yesterday / titleDay를 가져오던 방식을 뺏습니다.
- Date에 Days / DayFormat Enumerate를 추가하였고 static function을 통해 switch 로 dayString 정보를 가져올수 있게 변경했습니다.

volgaHwang committed:
refactor: BoxOfficeListCell StackView 네이밍 변경사항 반영

- leftStackView / rightStackView 는 너무 추상적인 네이밍이라 좀 더 포괄적인 내용을 담고 있는 네이밍으로 변경했습니다.

volgaHwang committed:
feat: Dynamic Type 적용을 위한 AttributedString 에 Computed Property 추가 반영

zdodev commented on BoxOffice/Scene/CellConfigurations/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,138 @@
+}
+
+extension BoxOfficeListCell {
+    func configure() {
```
뷰의 설정을 `configure` 에 작성하신 이유가 있을까요?
설정하는 구문을 뷰 초기화할 때 정의하는 것은 어떨까요?
이외에도 `configure` 메서드에서 많은 일을 처리하고 있는 것 같습니다. 이를 다른 메서드로 분리해서 정의해볼 수 있을까요?

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,132 @@
+import UIKit
+
+class BoxOfficeListViewController: UIViewController {
+    enum Section {
```
`internal` 접근 수준으로 정의하신 이유가 있을까요?

volgaHwang replied:
`Modern CollectionView` 에서 Layout 정의 하는 방식을 구현만하고 수정하지 않은 것 같습니다 ㅎㅎ;;

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,132 @@
+        case main
+    }
+    
+    private lazy var refresh: UIRefreshControl = {
```
`refresh` 프로퍼티 이름이 모호한 것 같습니다! 더 구체적인 이름을 사용할 수 있을까요?

volgaHwang replied:
`refreshBoxOfficeList`이 조금 더 구체적이라고 판단이 듭니다 ㅎㅎ

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,132 @@
+        
+        return refreshControl
+    }()
+    private lazy var collectionView: UICollectionView = {
+        let collectionView = UICollectionView(frame: .zero,
+                                              collectionViewLayout: self.createLayout())
```
해당 방법도 괜찮은 방법이라고 생각합니다!

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,132 @@
+        loadData()
+    }
+    
+    func loadData() {
```
`internal` 접근 수준으로 정의하신 이유가 있을까요? 다른 메서드도 확인해보면 좋을 것 같습니다!

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,132 @@
+    }
+    
+    @objc func refreshCollectionView() {
+        loadData()
+        refresh.endRefreshing()
```
`loadData` 메서드가 내부적으로 비동기로 동작하는데 그러면 데이터가 로드되기 전에 `endRefreshing` 메서드가 실행되지 않을까 싶습니다.
비동기 로직이 완료되고 나서 `endRefreshing` 메서드가 실행되도록 변경해보는 것은 어떨까요? 의견이 궁금합니다!

volgaHwang replied:
언급 주신대로 `loadData` 실행 요청 이후 `endRefreshing`이 되서 수행이 이상하게 됩니다 ㅎㅎ
클로져 사용해서 한번해보려고 합니다~

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,132 @@
+        title = Date.dayString(.yesterday, format: .yyyy_MM_dd)
+    }
+    
+    func configureHierarchy() {
```
`configureConstraints` 이름은 어떨까요?

zdodev commented on BoxOffice/Scene/CellConfigurations/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,138 @@
+class BoxOfficeListCell: UICollectionViewCell {
+    static let reuseIdentifier = "box-office-list-cell-reuseidentifier"
+    
+    let accessoryImageView = UIImageView()
```
`internal` 접근 수준으로 정의하신 이유가 있을까요? 다른 프로퍼티도 확인해보면 좋을 것 같습니다!

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+
+import UIKit
+
+struct BoxOfficeItem: Hashable {
```
`Cell` 에서 사용할 모델을 별도로 정의해보셨네요!👍

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+    let rank: String
+    let rankDescription: NSAttributedString
+    
+    private let identifier = UUID()
```
`identifier` 프로퍼티의 용도가 궁금합니다!

volgaHwang replied:
밑에 `hash()` 메소드를 작성했던 이유와 동일한 이유일 것 같습니다.
`Hashable` 프로토콜을 채택 할 경우, `hash()` 메소드가 아래 첨부한 사진같이 되어있어서 필요한 메소드로 알고있었습니다.
<img width="711" alt="스크린샷 2023-02-01 오전 1 07 55" src="https://user-images.githubusercontent.com/121410643/215815176-9e1a0ab5-aa32-442c-a2b2-610475a9fc67.png">

그런데, 공식문서를 다시 읽어보니, 
<img width="962" alt="스크린샷 2023-02-01 오전 1 08 47" src="https://user-images.githubusercontent.com/121410643/215815435-8c2d7cf9-50eb-4808-8770-26ff86370234.png">
`Struct` 의 프로퍼티들이 다 `Hashable`한 경우에 해당하므로 딱히 필요가 없다는 것을 확인했습니다 ㅎㅎ;

해당 부분 지워보겠습니다!

zdodev replied:
질문의 의도를 잘 파악해주셨습니다!🤗
공식문서의 추가내용 학습까지 좋습니다!👍

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+        
+        static func showCountInformation(_ todayShowCount: String,
+                                         _ totalShowCount: String) -> String {
+            return "오늘 \(totalShowCount) / 총 \(totalShowCount)"
```
오늘 카운트와 총 카운트가 동일합니다!

volgaHwang replied:
헉.. 고쳐야겠네요!

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+    
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(identifier)
+    }
```
`hash` 메서드를 구현하신 이유가 궁금합니다!

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+private extension String {
```
`private extension` 으로 한정해보셨네요!👍

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+}
+
+private extension String {
+    var thousandComma: String {
```
`thousandComma` 이름이 모호할 수 있을 것 같습니다. `locale` 또는 숫자 표기법과 관련된 이름으로 바꿔보는 것은 어떨까요?

volgaHwang replied:
보통 용어로 thousand separator로 표현을 하는듯 하여 해당 부분 네이밍 변경하겠습니다.

zdodev commented on BoxOffice/Entity/BoxOfficeItem.swift:
```diff
@@ -0,0 +1,115 @@
+            return ""
+        }
+        
+        let numberFormmater = NumberFormatter()
```
`NumberFormatter` 를 사용해보셨네요!👍

zdodev reviewed (changes requested):
안녕하세요. @volgaHwang 

이번 설 연휴는 아주 잘보냈습니다🤗
볼가도 즐겨운 연휴보내셨을까요~?

Step 3 구현 잘 진행해주셨습니다!

질문에 대한 제 의견을 적어보았습니다!

❓아직까지는? BoxOfficeListViewController에 코드가 감당 안될 정도로 많은편은 아니지만 조만간 더 많아질 것 같다는 생각이 드는 구조인데, 이 구조에 확신이 없는것같습니다. 혹시 더 컴팩트하게 줄일 방법이 있을까요?
- 현재도 잘 구현해주셨습니다!👍 점점 코드가 복잡해지고 많아질 것 같다는 생각만으로도 좋은 발전이라고 생각합니다! 다음 내용을 고민해보는 것은 어떨까요?
- 어떤 점에서 코드가 많다는 생각이 들었을까요?
- 앞으로 코드가 복잡해진다면 어떠한 점 때문에 복잡해질까요?
- 혹시 `ViewController` 가 많은 역할을 하고 있지는 않을까요?

추가로 기능 명세 중 다음 사항이 구현되지 않은 것 같습니다.
`처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.`
해당 기능을 구현해볼 수 있을까요?

다른 질문은 코멘트로 작성해보았습니다. 
다른 궁금증과 함께 확인해보면 좋을 것 같습니다!

volgaHwang committed:
refactor: BoxOfficeListViewController 변경사항 반영

- 접근제한자
    - `Enum Section` private
    - 메소드: `loadData`, `refreshCollectionView`, `createLayout`, `configureAttribute`, `configureConstraints`, `configureDataSource`, `updateUI` 접근제한자 `private` 추가
- 요구사항 반영(`UIActivityIndicatorView`)추가
- `loadData()` 메소드는 네트워크 동작이 끝나고 난 이후 `updateUI()` 호출과 `CollectionView`의 refreshing 그리고 처음 앱 시작 후 네트워킹 성공 후에 `activityView`를 꺼야할 의무가 있고, 해당 내용은 `escaping Closure`를 사용해서 비동기로 내용을 수행하게 했습니다.

volgaHwang committed:
refactor: BoxOfficeListCell 프로퍼티 접근제한자 추가 반영

volgaHwang committed:
refactor: BoxOfficeListCell 메소드 접근제한자 추가 + 기존 `configure`메소드 역할 분리

volgaHwang committed:
refactor: BoxOfficeListCell Property에 대한 직접 접근을 제한하고, fillContents를 사용해서 세팅할수 있게 구현했습니다.

volgaHwang committed:
refactor: BoxOfficeListViewController: cell filling 메소드 추가하여 반영

volgaHwang committed:
fix: BoxOfficeItem에 showCountInformation에서 잘못된 리턴값에 대한 수정사항 반영

volgaHwang committed:
refactor: BoxOfficeItem `Hashable` 프로토콜 채택에 불필요한 요소 삭제 및 String Extension Computed Property 변경사항 반영

volgaHwang commented:
@zdodev 

안녕하세요! 소대!

코멘트 주신 부분들에 대해 검토해보고 수정사항을 반영했습니다.

## 수정사항

- `TodayCount` / `TotalCount` 수정
- 외부에서 접근 할 필요 없는 객체 `property` 접근제한자 추가
- `Computed Property` 네이밍 변경
- `Cell` 의 `Configuration` 역할에 맡게 네이밍 분리
- `BoxOfficeItem` 구조체에서 `Hashable` 프로토콜 채택에 불필요한 `hash()`메소드 + `identifier` 프로퍼티 제거
- 미반영 요구사항 `처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.` 반영했습니다.
    - `loadData()` 메소드에 `escaping closure`로 `completion` 파라미터 추가하여 `endRefreshing`, `activityIndicator` 정지 시키게 사용

추가적으로 문의주셨던 `hash()`, `identifier` 프로퍼티를 사용했던 이유를 여쭤보셨는데 조사해보니 불필요한 이유를 찾아서 해당 내용을 대신 코멘트로 남겼습니다!

> 어떤 점에서 코드가 많다는 생각이 들었을까요?
- `configure` 나 특정 `behavior`에 대한 정의를 다 `ViewController`에 작성하다 보니 코드가 늘어나는거 아닌가 하는 생각이 들었습니다.
- 내용이 더 많아지게 될 경우, 어떤 코드를 작성했는지 좀 헷갈릴수 있겠다 라는 생각이 들었어요
> 앞으로 코드가 복잡해진다면 어떠한 점 때문에 복잡해질까요?
- 기능이 추가 될 경우, 해당 기능에 대해 메소드와 프로퍼티 가 `Massive`해져서 복잡해질 것 같아요. 위에 언급한것 과 같이 헷갈릴 것 같습니다..!
> 혹시 ViewController 가 많은 역할을 하고 있지는 않을까요?
- 맞습니다. ViewController가 많은 역할을 하고 있어서 해당 역할을 좀 어떻게 분산해야 할지에 대해 고민을 많이 해야할것 같습니다 ㅠㅠ

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,148 @@
+        case main
+    }
+    
+    private lazy var refreshBoxOfficeList: UIRefreshControl = {
```
좋습니다!👍

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -0,0 +1,148 @@
+    }
+    
+    @objc private func refreshCollectionView() {
+        loadData { [weak self] in
+            DispatchQueue.main.async {
+                self?.refreshBoxOfficeList.endRefreshing()
+            }
+        }
```
클로저를 사용하여 해결해보셨네요!👍

zdodev reviewed (approved):
안녕하세요. @volgaHwang !

피드백 반영된 부분 잘 확인하였습니다!👍

추가적으로, 복잡해지고 비대해지는 `ViewController` 문제를 해결하기 위한 많은 방법이 존재합니다.
`ViewController` 를 어떻게 하면 확장 가능하면서 가독성 있게 구현해나갈 수 있을지 먼저 고민해 보고 이를 해결하기 위한 방법들을 보면 더욱 좋지 않을까 싶습니다!

Step 3 진행하시느라 고생하셨습니다!👍
